src/LibraryManager/Json/LibraryStateToFileConverter.cs

```diff
@@ -22,22 +27,49 @@ public ILibraryInstallationState ConvertToLibraryInstallationState(LibraryInstal
             }
 
             string provider = string.IsNullOrEmpty(stateOnDisk.ProviderId) ? _defaultProvider : stateOnDisk.ProviderId;
-            string destination = string.IsNullOrEmpty(stateOnDisk.DestinationPath) ? _defaultDestination : stateOnDisk.DestinationPath;
+
+            (string name, string version) = LibraryIdToNameAndVersionConverter.Instance.GetLibraryNameAndVersion(stateOnDisk.LibraryId, provider);
+            string destination = string.IsNullOrEmpty(stateOnDisk.DestinationPath) ? ExpandDestination(_defaultDestination, name, version) : stateOnDisk.DestinationPath;
 
             var state = new LibraryInstallationState()
             {
+                Name = name,
+                Version = version,
                 IsUsingDefaultDestination = string.IsNullOrEmpty(stateOnDisk.DestinationPath),
                 IsUsingDefaultProvider = string.IsNullOrEmpty(stateOnDisk.ProviderId),
                 ProviderId = provider,
                 DestinationPath = destination,
                 Files = stateOnDisk.Files
             };
 
-            (state.Name, state.Version) = LibraryIdToNameAndVersionConverter.Instance.GetLibraryNameAndVersion(stateOnDisk.LibraryId, provider);
-
             return state;
         }
 
+        /// <summary>
+        /// Expands [Name] and [Version] tokens in the DefaultDestination
+        /// </summary>
+        /// <param name="destination">The default destination string</param>
+        /// <param name="name">Package name</param>
+        /// <param name="version">Package version</param>
+        /// <returns></returns>
+        [SuppressMessage("Globalization", "CA1307:Specify StringComparison for clarity", Justification = "Not available on net481, not needed here (caseless)")]
+        private string ExpandDestination(string destination, string name, string version)
+        {
+            if (!destination.Contains("["))
+            {
+                return destination;
+            }
+
+            // if the name contains a slash (either filesystem or scoped packages),
+            // trim that and only take the last segment.
+            int cutIndex = name.LastIndexOfAny(['/', '\\']);
+
+            StringBuilder stringBuilder = new StringBuilder(destination);
+            stringBuilder.Replace("[Name]", cutIndex == -1 ? name : name.Substring(cutIndex + 1));
```

Comment:
Possibly a non-real case, but what if name contains [Version]? That would cause the next line to do an extra/unexpected replacement.

Reply:
At least as far as NPM is concerned, I don't believe that is a valid name (only allows alphanumeric, underscores, and hyphens).  Since that's the source for the non-filesystem providers (jsdelivr and unpkg directly use NPM, cdnjs is indirectly curated from NPM), it would be an edge case for filesystem only.  I thought about making this logic provider-specific, but that was a lot more complicated than doing it here centrally.

I did briefly consider using a different token that wouldn't be allowed in file paths (implicitly compatible with all providers since they all come down to file names), but on Linux the only character is `/`.  I'm definitely open another delineator if you have a preference, but `[Version]` felt most natural (next on my list was something like `{{Version}}`); it should be a pretty trivial change.